File: alternative/medium/word_break.py
from typing import List

def wordBreak(s: str, wordDict: List[str]) -> bool:
    # Convert the wordDict to a set for faster lookup
    wordSet = set(wordDict)
    # Create a dp array of size s + 1
    dp = [False] * (len(s) + 1)
    # Set the first element of the dp array to True
    dp[0] = True

    # Iterate through the dp array Bottom Up
    for i in range(1, len(dp)):
        # Iterate through the dp array again
        for j in range(i):
            # If the dp[j] is True and the substring from j to i is in the wordSet
            if dp[j] and s[j:i] in wordSet:
                # Set the dp[i] to True
                dp[i] = True
                # Break out of the loop
                break
        
    # Return the last element of the dp array
    return dp[-1]

assert wordBreak("leetcode", ["leet", "code"]) == True
assert wordBreak("leetleet", ["leet", "code"]) == True
assert wordBreak("applepenapple", ["apple", "pen"]) == True
assert wordBreak("catsandog", ["cats", "dog", "sand", "and", "cat"]) == False        

File: src/medium/word_break.rs
#![allow(dead_code)]
pub fn word_break(s: String, word_dict: Vec<String>) -> bool {
    let word_set: std::collections::HashSet<_> = word_dict.into_iter().collect();
    let mut dp = vec![false; s.len() + 1];
    dp[0] = true;

    for i in 1..=s.len() {
        for j in 0..i {
            if dp[j] && word_set.contains(&s[j..i]) {
                dp[i] = true;
                break;
            }
        }
    }

    dp[s.len()]
}

#[test]
fn test_word_break() {
    assert_eq!(
        word_break(
            "leetcode".to_string(),
            vec!["leet".to_string(), "code".to_string()]
        ),
        true
    );
    assert_eq!(
        word_break(
            "applepenapple".to_string(),
            vec!["apple".to_string(), "pen".to_string()]
        ),
        true
    );
    assert_eq!(
        word_break(
            "catsandog".to_string(),
            vec![
                "cats".to_string(),
                "dog".to_string(),
                "sand".to_string(),
                "and".to_string(),
                "cat".to_string()
            ]
        ),
        false
    );
}
